### DISCUSSION TITLE
Question: Unable to get initial data using subscription

### DISCUSSION THREAD

**Discussion by adityapant1286** [enisdenjo/graphql-ws#408] [Q&A] [answered]
I am fairly new to using graphql-ws, so forgive me if this is a naive question or mistake from my side. 

## Libraries used:
**Backend**
- graphql-yoga
- ws
- graphql-ws

**Frontend**
- solid-js
- wonka

My understanding, there is something I am missing in the schema definition which is not showing the todo list when tried accessing _Yoga Graphiql explorer._
The documentation mentioned the schema as a parameter. The schema definition required for this library is missing. Need some inputs on what am I missing.

## Screenshot
![image](https://user-images.githubusercontent.com/13958935/193436344-df16ecc7-30a0-45ef-a577-fc4b8f651ca9.png)

## Code Snippets

***Schema definition***
```
import { createPubSub } from 'graphql-yoga';
import { Todo } from "./types";

let todos = [
    {
        id: "1",
        title: "Learn GraphQL + Solidjs",
        completed: false
    }
];

// channel
const TODOS_CHANNEL = "TODOS_CHANNEL";

// pubsub
const pubSub = createPubSub();

const publishToChannel = (data: any) => pubSub.publish(TODOS_CHANNEL, data);

// Type def
const typeDefs = [`
    type Todo {
        id: ID!
        title: String!
        completed: Boolean!
    }

    type Query {
        getTodos: [Todo]!
    }

    type Mutation {
        addTodo(title: String!): Todo!
    }

    type Subscription {
        todos: [Todo!]
    }
`];

// Resolvers
const resolvers = {
    Query: {
        getTodos: () => todos
    },
    Mutation: {
        addTodo: (_: unknown, { title }: Todo) => {
            const newTodo = {
                id: "" + (todos.length + 1),
                title,
                completed: false
            };
            todos.push(newTodo);
            publishToChannel({ todos });
            return newTodo;
        },
    Subscription: {
        todos: {
            subscribe: () => {
                const res = pubSub.subscribe(TODOS_CHANNEL);
                publishToChannel({ todos });
                return res;
            }
        },
    },
};

export const schema = {
    resolvers,
    typeDefs
};
```

***Server backend***
```
import { createServer } from "graphql-yoga";
import { WebSocketServer } from "ws";
import { useServer } from "graphql-ws/lib/use/ws";

import { schema } from "./src/schema";
import { execute, ExecutionArgs, subscribe } from "graphql";

async function main() {
    const yogaApp = createServer({
        schema,
        graphiql: {
            subscriptionsProtocol: 'WS', // use WebSockets instead of SSE
        },
    });

    const server = await yogaApp.start();
    const wsServer = new WebSocketServer({
        server,
        path: yogaApp.getAddressInfo().endpoint
    });

    type EnvelopedExecutionArgs = ExecutionArgs & {
        rootValue: {
            execute: typeof execute;
            subscribe: typeof subscribe;
        };
    };

    useServer(
        {
            execute: (args: any) => (args as EnvelopedExecutionArgs).rootValue.execute(args),
            subscribe: (args: any) => (args as EnvelopedExecutionArgs).rootValue.subscribe(args),
            onSubscribe: async (ctx, msg) => {
                const { schema, execute, subscribe, contextFactory, parse, validate } =
                    yogaApp.getEnveloped(ctx);

                const args: EnvelopedExecutionArgs = {
                    schema,
                    operationName: msg.payload.operationName,
                    document: parse(msg.payload.query),
                    variableValues: msg.payload.variables,
                    contextValue: await contextFactory(),
                    rootValue: {
                        execute,
                        subscribe,
                    },
                };

                const errors = validate(args.schema, args.document);
                if (errors.length) return errors;
                return args;
            },
        },
        wsServer,
    );

}

main().catch((e) => {
    console.error(e);
    process.exit(1);
});
```

## Expected Behaviour
It should show the initial todo item as soon as it has subscribed and published in schema definition.
_Todo item - declared in schema_
```
    {
        id: "1",
        title: "Learn GraphQL + Solidjs",
        completed: false
    }
```

**Actual Behaviour**
It does not show/return initial todo item.
After adding a new todo (using `addTodo`) it shows two todo items. So I believe it is unable to return initial whenever refresh the browser or run subscribe on _Yoga Graphiql explorer_.

**Comment by enisdenjo** [comment, marked as answer]
This seems like a question about GraphQL subscriptions in general or `graphql-yoga`, rather than `graphql-ws`.

Are you sure that `createPubSub` has a queue? Meaning that if something gets published _before_ anyone is subscribed, it would get queued up and flushed when someone finally subscribes?

**Reply by adityapant1286** [reply]
Hi @enisdenjo, I think you're right. Unfortunately, there are only two methods in `PubSub`. It is a bit hard to know whether calling `.publish()` immediately after `.subscribe()` put the data in a queue. I will reach out to `graphql-yoga` team for more information. 
Thanks for your time and comments.

**Reply by enisdenjo** [reply]
I don't think the pubsub has a queue, `.publish()` only dispatches events to currently active subscribers. A queue is not easy to implement as something offered out-of-the-box.

I'd recommend you making one yourself. To make it efficient, you'd have to implement the pubsub yourself too - but that is not too complicated. Just be wary to empty the queue every once in a while to avoid memory leaks.

**Reply by enisdenjo** [reply]
Of course, you can also just make sure the app subscribes _first_, before creating any items. 🤷‍♂️
